sensorserviceutils/sensorserviceutils.go

```diff
@@ -0,0 +1,384 @@
+		`(?P<sensorNumber>0x?[0-9A-Fa-f]+)` + `\s*samplingPeriod=\s*` +
+		`(?P<samplingPeriod>\d+)us` + `\s*maxReportLatency=\s*` + `(?P<batchingPeriod>\d+)us`)
+
+	// removeRegistrationOldRE is a regular expression to match the log that removes subscription in
+	// the sensorservice dump in the bugreport starting from MNC or before.
+	removeRegistrationOldRE = regexp.MustCompile(`(?P<timeStamp>\d+\:\d+\:\d+)` + `\s*` +
+		`de\-activated` + `\s*package=\s*` + `(?P<packageName>[^(]+)` + `\s*handle=\s*` +
+		`(?P<sensorNumber>0x?[0-9A-Fa-f]+)`)
+
+	// timeLayoutRE is a regular expression to match the timestamp with dates that may probably
+	// show up in the bugreport.
+	timeLayoutRE = regexp.MustCompile(`^(?P<month>\d+)-(?P<day>\d+)`)
+)
+
+const (
+	sensorRegisDesc = "Sensorservice Registration"
+	activeConnDesc  = "Active Connection"
+	completeStr     = "complete"
+	activeStr       = "active"
+	errorStr        = "error"
+	// unknownTime is used when the start or end time of an event is unknown.
+	// This is not zero as csv.AddEntryWithOpt ignores events with a zero time.
+	unknownTime = -1
+)
+
+// OutputData contains information for active connection and previous registration history collected
+// in the sensorservice dump.
+type OutputData struct {
+	CSV         string
+	ActiveConns []*acpb.ActiveConn
+	ParsingErrs []error
+	SensorErrs  []error
+}
+
+// SubscriptionInfo contains information about one subscription event of a sensor to an application.
+// For NRD42 and onwards Android versions: easch subscription event is captured by the + statement
+// that adds the subscription and the - statement that removes the subscription .
+// For MNC or before: each subscription event is captured by the activated/de-activated statments.
+type SubscriptionInfo struct {
+	StartMs, EndMs int64
+	SensorNumber   int32
+	PackageName    string
+	SamplingPeriod int32
+	BatchingPeroid int32
+}
+
+type parser struct {
+	// referenceYear is the year extracted from the dumpstate line in a bugreport.
+	// Previous Registration lines don't contain a year in the date string, so we use this
+	// to reconstruct the full timestamp.
+	referenceYear int
+
+	// referenceMonth is the month extracted from the dumpstate line in a bugreport.
+	// Since a bugreport may span over a year boundary, we use the month to check whether the
+	// year for the event needs to be decremented or incremented.
+	referenceMonth time.Month
+
+	// referenceDay is the month extracted from the dumpstate line in a bugreport.
+	referenceDay int
+
+	// loc is the location parsed from timezone information in the bugreport.
+	// The previous registration is in the user's local timezone which we need to convert to UTC time.
+	loc *time.Location
+
+	lines []string
+	idx   int
+
+	buf         *bytes.Buffer
+	csvState    *csv.State
+	parsingErrs []error
+	sensorErrs  []error
+
+	// sensors is a map from sensor number to the relevant sensor's information.
```

Comment:
By sensor number, do you mean the handle?

Reply:
Yes. I am using the phrase "sensor number" across all files to refer to the handle ID. It will be clearer if I keep the notations the same.